Metric Definitions page

ivan-kishko opened the pull request:
## Description of the change

> New page for metric definitions
- breakdowns list with tooltip
- modal for each breakdown/metric total
- save logic changed, now all (included/excluded and contexts) save at once on save click
- P.S. it turned out that on init definitions context was not initializing with value but only label for some reason, fixed that

## Type of change
Type: Non-breaking refactor 

> All pull requests must have at least one of the following labels applied (otherwise the PR will fail):

| Label                       	| Description                                                                                               	|
|-----------------------------	|-----------------------------------------------------------------------------------------------------------	|
| Type: Bug                   	| non-breaking change that fixes an issue                                                                   	|
| Type: Feature               	| non-breaking change that adds functionality                                                               	|
| Type: Breaking Change       	| fix or feature that would cause existing functionality to not work as expected                            	|
| Type: Non-breaking refactor 	| change addresses some tech debt item or prepares for a later change, but does not change functionality    	|
| Type: Configuration Change  	| adjusts configuration to achieve some end related to functionality, development, performance, or security 	|
| Type: Dependency Upgrade      | upgrades a project dependency - these changes are not included in release notes                             |

## Related issues

Closes #562 (Last part)

## Checklists

### Development

**This box MUST be checked by the submitter prior to merging**:
- [ ] **Double- and triple-checked that there is no Personally Identifiable Information (PII) being mistakenly added in this pull request**

These boxes should be checked by the submitter prior to merging:
- [ ] Tests have been written to cover the code changed/added as part of this pull request

### Code review

These boxes should be checked by reviewers prior to merging:

- [ ] This pull request has a descriptive title and information useful to a reviewer
- [ ] This pull request has been moved out of a Draft state, has no "Work In Progress" label, and has assigned reviewers
- [ ] Potential security implications or infrastructural changes have been considered, if relevant

lilidworkin commented:
@ivan-kishko this looks fantastic!

Two small things:
* I don't see the "Configuration Required" indicator state anywhere (the orange one) -- what's the logic for when that shows up?
* One small edge case with the "Available for Publishing" indicator that I just thought of -- if a breakdown is not selected, then we shouldn't require the user to select a checkmark for it in order for this indicator to appear. For instance, in the video attached, none of the breakdowns are selected, and I go into "Define Metrics", and I select a checkmark for the top-level metric -- then it should say "Available for Publishing". If I turn on a breakdown, then I should be required to select a checkmark for that breakdown. But if it's not turned on, then I shouldn't be. Let me know if that makes sense! Also cc @mxosman , curious if this makes sense to you / seems like the right behavior. I'm not 100% sure, haha.

https://user-images.githubusercontent.com/99899591/234966310-04774c4b-fcc5-4f50-afc7-f01386c2e148.mov

mxosman commented:
> * One small edge case with the "Available for Publishing" indicator that I just thought of -- if a breakdown is not selected, then we shouldn't require the user to select a checkmark for it in order for this indicator to appear. For instance, in the video attached, none of the breakdowns are selected, and I go into "Define Metrics", and I select a checkmark for the top-level metric -- then it should say "Available for Publishing". If I turn on a breakdown, then I should be required to select a checkmark for that breakdown. But if it's not turned on, then I shouldn't be. Let me know if that makes sense! Also cc @mxosman , curious if this makes sense to you / seems like the right behavior. I'm not 100% sure, haha.

I was recording a similar video actually when I was playing around with it -- what you explained makes sense and is 100% the correct behavior. 

Thank you for working on a fix, Ivan!

This overall feels really great! Aside from the above, one other thing that stood out to me were the tooltips. They feel like they get in the way a little bit and persist even when the mouse is outside of the box (if it's still hovering inside the tooltip). Is there any way we can have the tooltip appear on the bottom left or right of the dimension box? Something like below -- but I'm not sure if that too feels like it gets in the way. Any thoughts @lilidworkin/Ivan?

<img width="1001" alt="Screenshot 2023-04-27 at 4 52 57 PM" src="https://user-images.githubusercontent.com/59492998/234999194-84940546-909b-476a-a1cf-2d83be28f479.png">

mxosman commented on publisher/src/components/MetricSettings/MetricDefinitions.styled.tsx:
```diff
@@ -0,0 +1,101 @@
+
+export const HeaderLabel = styled(MetricAvailability.HeaderLabel)``;
+
+export const Description = styled(MetricAvailability.Description)`
```
I really do like this way of importing all styled components into a variable - makes it super clear where you're inheriting from.

mxosman commented on publisher/src/components/MetricSettings/DefinitionModalForm.tsx:
```diff
@@ -0,0 +1,441 @@
+      !activeSettingsKeys?.length && activeDimensionKey && !dimensionContextsMap
+    );
+
+  const hasNoSettings =
```
nit: this var name is confusing with the one in 104. maybe `hasNoSettingsAndNoContext` or something similar?

mxosman commented on publisher/src/components/MetricSettings/DefinitionModalForm.tsx:
```diff
@@ -0,0 +1,441 @@
+    undefined;
+
+  const [currentSettings, setCurrentSettings] = useState(
+    activeSettingsKeys?.reduce((acc, includesExcludesKey) => {
+      const currentIncludesExcludes =
+        activeDisaggregationKey && activeDimensionKey
+          ? dimensionDefinitionSettings[systemMetricKey][
+              activeDisaggregationKey
+            ][activeDimensionKey][includesExcludesKey]
+          : metricDefinitionSettings[systemMetricKey][includesExcludesKey];
+
+      return {
+        ...acc,
+        [includesExcludesKey]: currentIncludesExcludes,
+      };
+    }, {} as { [includesExcludesKey: string]: { description?: string; settings: { [settingKey: string]: Partial<MetricConfigurationSettings> } } })
```
Can we extract this out into a variable like `initialSettings` and pass that into the `useState`?

ivan-kishko replied:
[50a7d0a](https://github.com/Recidiviz/justice-counts/pull/593/commits/50a7d0ad1d03eeba6ece0834339de3c74f6e9bd1)

mxosman commented on publisher/src/components/MetricSettings/DefinitionModalForm.tsx:
```diff
@@ -0,0 +1,441 @@
+    Object.keys(metricDefinitionSettings[systemMetricKey]);
+
+  const dimensionDefinitionSettingsKeys = (activeDisaggregationKey &&
+    activeDimensionKey &&
```
Can we move L87-88 up so we can use that to save us from writing out `activeDisaggregationKey && activeDimensionKey`?

ivan-kishko replied:
[50a7d0a](https://github.com/Recidiviz/justice-counts/pull/593/commits/50a7d0ad1d03eeba6ece0834339de3c74f6e9bd1)

mxosman commented on publisher/src/components/MetricSettings/DefinitionModalForm.tsx:
```diff
@@ -0,0 +1,441 @@
+
+  const currentDimension =
+    (activeDisaggregationKey &&
+      activeDimensionKey &&
```
nit: could use `hasActiveDisaggregationAndDimensionKey`

mxosman commented on publisher/src/components/MetricSettings/DefinitionModalForm.tsx:
```diff
@@ -0,0 +1,441 @@
+          {} as {
+            [contextKey: string]: { label: string; value: string };
+          }
+        )
```
Same as above with extracting this reducer logic into a variable. Also, the explicit types are becoming a bit unwieldy - can we extract those out and define them elsewhere too?

mxosman commented on publisher/src/components/MetricSettings/DefinitionModalForm.tsx:
```diff
@@ -0,0 +1,441 @@
+            };
+          };
+        }
+      )
```
Same here with this - it feels a bit hard to read and I wonder if there's a better way to express this without nesting a reducer.

ivan-kishko replied:
Well it is yes, but we have to loop through includesExcludes keys and then we have to loop through all settings of every single includesExcludes key. And unlike previous iteration of "return to default" where we had to just update each includeExclude here we have to return an object in form of currentSettings useState since we are not saving anything until user hits Save button. There was nested looping as well just with forEach, but it was like that cause all changes were saved right away

mxosman commented on publisher/src/components/MetricSettings/MetricConfig.tsx:
```diff
@@ -52,6 +56,56 @@ function MetricConfig() {
+        return { ...acc, ...dimensionDefinitionSettings[systemMetricKey][key] };
+      },
+      {} as {
+        [dimensionKey: string]: {
+          [includesExcludesKey: string]: {
+            description?: string;
+            settings: {
+              [settingKey: string]: Partial<MetricConfigurationSettings>;
+            };
+          };
+        };
```
nit: let's create a type and pass it in here

ivan-kishko replied:
[50a7d0a](https://github.com/Recidiviz/justice-counts/pull/593/commits/50a7d0ad1d03eeba6ece0834339de3c74f6e9bd1)

mxosman commented on publisher/src/components/MetricSettings/DefinitionModalForm.tsx:
```diff
@@ -0,0 +1,441 @@
+      const settingsToSave: MetricSettings["settings"] = [];
+      const contextsToSave: MetricSettings["contexts"] = [];
+
+      if (activeDisaggregationKey && activeDimensionKey) {
```
Can we add comments to clarify what's happening within & outside this conditional?

ivan-kishko replied:
[50a7d0a](https://github.com/Recidiviz/justice-counts/pull/593/commits/50a7d0ad1d03eeba6ece0834339de3c74f6e9bd1)

mxosman commented:
The behavior with the Race/Ethnicities looks exactly right - there wouldn't be any includes/excludes for them anyway - so it makes sense to just have them live in the Breakdown section. For how it affects the "Availability for publishing" - it should affect it the same way other breakdowns without definitions do - turning it on/off is sufficient to say this breakdown is complete.

lilidworkin commented:
> For how it affects the "Availability for publishing" - it should affect it the same way other breakdowns without definitions do - turning it on/off is sufficient to say this breakdown is complete.

Yes, what @mxosman says here is right! I think this might be how you have it already @ivan-kishko , but basically, if a breakdown doesn't have definitions, it shouldn't affect "Available for publishing". As long as all breakdowns that DO have definitions have at least one definition selected, the "Available for publishing" indicator should still show up. LMK if that makes sense!

ivan-kishko commented:
@lilidworkin @mxosman Hello! Please check out this [d543211](https://github.com/Recidiviz/justice-counts/pull/593/commits/d543211378478a8d26db251e1cc9aaecc93f555e) commit foe changes regarding available dimensions check. Also added a check for at least one race selected (cause if no races selected then race/ethnicities is disabled, I guess that is what you were talking about right?)

mxosman reviewed (approved):
> @lilidworkin @mxosman Hello! Please check out this [fdbcf06](https://github.com/Recidiviz/justice-counts/pull/593/commits/fdbcf0620c23be9f2da8db93ad5fad54b22db69f) commit for changes regarding available dimensions check.

The behavior works exactly as expected! Thank you, Ivan!

My one design nit here is I wish the color/style of `Available for data upload` and `Available for publishing` were slightly different because it's not immediately noticeable when it changes.


> @mxosman I just changed tooltip width and moved it a bit more left [ffeace9](https://github.com/Recidiviz/justice-counts/pull/593/commits/ffeace9b8ecec206c7cedf14fa1401eb1a645610). If we do like you suggest then next 1 or 2 dimension names are covered with it so I just decided to make it take less place in x axis so everything (name and right "view/edit" is visible enough) LMK if it looks better

Thank you, Ivan! I like that you can see the dimension name and the view/edit with your changes. 

Nothing for you to do - but another design nit would be to possibly lose the tooltip in favor of a hover explanation underneath the "Click into each of the datapoints you are sharing to tell us what events, populations, etc. are included in that datapoint. Learn more in the Technical Implementation Guide.". 

I'll bring up both nits in the channel - but everything you did here looks great to me!

lilidworkin reviewed (approved):
Looks great @ivan-kishko ! Two tiny things and then you can merge:

When there are no definitions, can we remove the text at the top that says "Indicate which ..." and just go straight to the text that says "Please describe what data is being included in this breakdown."?

<img width="600" alt="Screenshot 2023-05-01 at 11 41 08 AM" src="https://user-images.githubusercontent.com/99899591/235480202-4419ba47-6d42-47fb-a1be-8e3d58dc798e.png">

When there are no definitions or contexts, and the dialog is empty, is it possible to just not have it show up at all? Like have "Unknown Offenses" not be clickable and take away the "View/Edit" link? If this is tricky, don't worry about it. In that case, let's just change the text to say "There are no definitions to configure for this breakdown." (take out the word "yet").

<img width="600" alt="Screenshot 2023-05-01 at 11 41 17 AM" src="https://user-images.githubusercontent.com/99899591/235480224-2bc23ce9-cfe1-42b6-b119-faea0e125dad.png">

Thank you!!

ivan-kishko commented:
> Looks great @ivan-kishko ! Two tiny things and then you can merge:
> 
> When there are no definitions, can we remove the text at the top that says "Indicate which ..." and just go straight to the text that says "Please describe what data is being included in this breakdown."?
> 
> <img alt="Screenshot 2023-05-01 at 11 41 08 AM" width="600" src="https://user-images.githubusercontent.com/99899591/235480202-4419ba47-6d42-47fb-a1be-8e3d58dc798e.png">
> 
> When there are no definitions or contexts, and the dialog is empty, is it possible to just not have it show up at all? Like have "Unknown Offenses" not be clickable and take away the "View/Edit" link? If this is tricky, don't worry about it. In that case, let's just change the text to say "There are no definitions to configure for this breakdown." (take out the word "yet").
> 
> <img alt="Screenshot 2023-05-01 at 11 41 17 AM" width="600" src="https://user-images.githubusercontent.com/99899591/235480224-2bc23ce9-cfe1-42b6-b119-faea0e125dad.png">
> 
> Thank you!!

Hello @lilidworkin. For the first I removed description. About second - well it is not tricky but I have to either copy some logic from modal or do all the logic outside modal. I went with 2nd option where I just removed "yet" and change button to be red with label "Close" so it differs from modal with available settings. When or If you decide later to make it another way just make a task for me and I fix it! For now I am gonna merge.

lilidworkin commented:
Great thanks @ivan-kishko , that solution sounds great to me!
